include/llvm-dialects/Dialect/OpSet.h

```diff
@@ -0,0 +1,267 @@
+  bool containsCoreOpcode(unsigned coreOpcode) const {
+    return m_coreOpcodes.contains(coreOpcode);
+  }
+
+  // Checks if a given intrinsic ID is stored in the set.
+  bool containsIntrinsicID(unsigned intrinsicID) const {
+    return m_intrinsicIDs.contains(intrinsicID);
+  }
+
+  // Checks if a given dialect operation is stored in the set.
+  template <typename OpT> bool contains() const {
+    static OpDescription desc = OpDescription::get<OpT>();
+    return contains(desc);
+  }
+
+  // Checks if a given OpDescription is stored in the set.
+  bool contains(const OpDescription &desc) const {
+    if (desc.isCoreOp()) {
+      for (const unsigned op : desc.getOpcodes()) {
+        if (containsCoreOpcode(op))
+          return true;
+      }
+    }
+
+    if (desc.isIntrinsic()) {
+      for (const unsigned op : desc.getOpcodes()) {
+        if (containsIntrinsicID(op))
+          return true;
+      }
+    }
+
+    return isMatchingDialectOp(desc.getMnemonic());
+  }
+
+  // -------------------------------------------------------------
+  // Match checks to determine if the type of a given instruction is stored in
+  // the OpSet and matches against a given type.
+  // -------------------------------------------------------------
+  // Checks if `inst` is of type OpT, if it is stored in the set.
+  template <typename OpT> bool matches(const Instruction &inst) const {
+    static OpDescription desc = OpDescription::get<OpT>();
+    return matches(desc, inst);
+  }
+
+  // Checks if `inst` is described by `desc`, if it is stored in the
+  // set.
+  bool matches(const OpDescription &desc, const Instruction &inst) const {
+    if (isa<IntrinsicInst>(inst) || isa<CallInst>(inst)) {
+      return matches(desc, *cast<CallInst>(inst).getCalledFunction());
+    }
+
+    if (desc.isCoreOp()) {
+      for (const unsigned op : desc.getOpcodes()) {
+        if (containsCoreOpcode(op) && desc.matchInstruction(inst))
+          return true;
+      }
+    }
+
+    return false;
+  }
+
+  // Checks if `func` is of type OpT, if it is stored in the set.
+  template <typename OpT> bool matches(const Function &func) const {
+    static OpDescription desc = OpDescription::get<OpT>();
+    return matches(desc, func);
+  }
+
+  // Checks if `func` is described by `desc`, if it is stored in the
+  // set.
+  bool matches(const OpDescription &desc, const Function &func) const {
+    if (desc.isIntrinsic()) {
+      for (const unsigned op : desc.getOpcodes()) {
+        if (containsIntrinsicID(op) && desc.matchDeclaration(func))
+          return true;
+      }
+    }
+
+    return contains(desc) && desc.matchDeclaration(func);
+  }
+
+  // -------------------------------------------------------------
+  // Membership tests check whether a given instruction or function is
+  // stored in the OpSet.
+  // -------------------------------------------------------------
+
+  // Checks if `inst` belongs to the OpSet.
+  bool isMember(const Instruction &Inst) const {
```

Comment:
Why not make this an overload of `contains`. That feels a bit more intuitive to me.